File: bench/regexredux/regexredux.go
/* The Computer Language Benchmarks Game
 * http://benchmarksgame.alioth.debian.org/
 *
 * regex-dna program contributed by The Go Authors.
 * converted from regex-dna program
 */

package main

import (
   "fmt"
   "io/ioutil"
   "os"
   "runtime"
   "regexp"
)

var variants = []string{
   "agggtaaa|tttaccct",
   "[cgt]gggtaaa|tttaccc[acg]",
   "a[act]ggtaaa|tttacc[agt]t",
   "ag[act]gtaaa|tttac[agt]ct",
   "agg[act]taaa|ttta[agt]cct",
   "aggg[acg]aaa|ttt[cgt]ccct",
   "agggt[cgt]aa|tt[acg]accct",
   "agggta[cgt]a|t[acg]taccct",
   "agggtaa[cgt]|[acg]ttaccct",
}

type Subst struct {
   pat, repl string
}

var substs = []Subst{
   Subst{"tHa[Nt]", "<4>"},
   Subst{"aND|caN|Ha[DS]|WaS", "<3>"},
   Subst{"a[NSt]|BY", "<2>"},
   Subst{"<[^>]*>", "|"},
   Subst{"\\|[^|][^|]*\\|", "-"},
}

func countMatches(pat string, bytes []byte) int {
   re := regexp.MustCompile(pat)
   n := 0
   for {
      e := re.FindIndex(bytes)
      if e == nil {
         break
      }
      n++
      bytes = bytes[e[1]:]
   }
   return n
}

func main() {
   runtime.GOMAXPROCS(4)
   bytes, err := ioutil.ReadFile("/dev/stdin")
   if err != nil {
      fmt.Fprintf(os.Stderr, "can't read input: %s\n", err)
      os.Exit(2)
   }
   ilen := len(bytes)
   // Delete the comment lines and newlines
   bytes = regexp.MustCompile("(>[^\n]+)?\n").ReplaceAll(bytes, []byte{})
   clen := len(bytes)

   mresults := make([]chan int, len(variants))
   for i, s := range variants {
      ch := make(chan int)
      mresults[i] = ch
      go func(ss string) {
         ch <- countMatches(ss, bytes)
      }(s)
   }

   lenresult := make(chan int)
   bb := bytes
   go func() {
      for _, sub := range substs {
         bb = regexp.MustCompile(sub.pat).ReplaceAll(bb, []byte(sub.repl))
      }
      lenresult <- len(bb)
   }()

   for i, s := range variants {
      fmt.Printf("%s %d\n", s, <-mresults[i])
   }
   fmt.Printf("\n%d\n%d\n%d\n", ilen, clen, <-lenresult)
}

File: bench/regexredux/regexredux.swift
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// contributed by Francois Green


import Foundation
import Dispatch

let input = FileHandle.standardInput.readDataToEndOfFile()

var sequence = String(data: input, encoding: .utf8)!

let inputLength = input.count

let regex: (String) -> NSRegularExpression = { pattern in
  return try! NSRegularExpression(pattern: pattern, options: [])
}

sequence = regex(">[^\n]*\n|\n").stringByReplacingMatches(in: sequence, options: [], range: NSRange(location: 0, length: inputLength), withTemplate: "")

let codeLength = sequence.utf8.count

var resultLength: Int?

let group = DispatchGroup()

DispatchQueue.global(qos: .background).async {
  group.enter()
  resultLength = [
    (regex: "tHa[Nt]",            replacement: "<4>"),
    (regex: "aND|caN|Ha[DS]|WaS", replacement: "<3>"),
    (regex: "a[NSt]|BY",          replacement: "<2>"),
    (regex: "<[^>]*>",            replacement: "|"),
    (regex: "\\|[^|][^|]*\\|",    replacement: "-")
  ].reduce(sequence) { buffer, iub in
    return regex(iub.regex).stringByReplacingMatches(in: buffer, options: [], range: NSRange(location: 0, length: buffer.utf16.count), withTemplate: iub.replacement)
  }.utf8.count
  group.leave()
}

[
  "agggtaaa|tttaccct",
  "[cgt]gggtaaa|tttaccc[acg]",
  "a[act]ggtaaa|tttacc[agt]t",
  "ag[act]gtaaa|tttac[agt]ct",
  "agg[act]taaa|ttta[agt]cct",
  "aggg[acg]aaa|ttt[cgt]ccct",
  "agggt[cgt]aa|tt[acg]accct",
  "agggta[cgt]a|t[acg]taccct",
  "agggtaa[cgt]|[acg]ttaccct"
].forEach { variant in
  print(variant, regex(variant).numberOfMatches(in: sequence, options: [], range: NSRange(location: 0, length: sequence.utf8.count)))
}

group.wait()
print("", inputLength, codeLength, resultLength!, separator: "\n")
